src/main/java/cms/gov/madie/measure/services/BundleService.java

```diff
@@ -0,0 +1,60 @@
+package cms.gov.madie.measure.services;
+
+import cms.gov.madie.measure.exceptions.BundleOperationException;
+import cms.gov.madie.measure.exceptions.CqlElmTranslationErrorException;
+import cms.gov.madie.measure.exceptions.CqlElmTranslationServiceException;
+import cms.gov.madie.measure.exceptions.InvalidResourceBundleStateException;
+import com.nimbusds.oauth2.sdk.util.StringUtils;
+import gov.cms.madie.models.measure.ElmJson;
+import gov.cms.madie.models.measure.Measure;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+@Slf4j
+@Service
+@AllArgsConstructor
+public class BundleService {
+
+  private final FhirServicesClient fhirServicesClient;
+
+  private final ElmTranslatorClient elmTranslatorClient;
+
+  public String bundleMeasure(Measure measure, String accessToken) {
+    if (measure == null) {
+      return null;
+    }
+
+    if (measure.isCqlErrors()) {
+      throw new InvalidResourceBundleStateException(
+          "Measure", measure.getId(), "since CQL errors exist.");
+    }
+
+    if (CollectionUtils.isEmpty(measure.getGroups())) {
+      throw new InvalidResourceBundleStateException(
+          "Measure", measure.getId(), "since there are no associated measure groups.");
```

Comment:
As Ashok pointed out to me yesterday, this is good chance to update "measure groups" to "Population Criteria".